jetty-server/src/main/java/org/eclipse/jetty/server/HttpInputOverHTTP.java

```diff
@@ -18,18 +18,37 @@
 
 package org.eclipse.jetty.server;
 
-import java.io.IOException;
-
-public class HttpInputOverHTTP extends HttpInput
+// tests used: RequestTest, PartialRFC2616Test, AsyncRequestReadTest, AsyncIOServletTest, GzipHandlerTest
+public class HttpInputOverHTTP extends AbstractHttpInput
 {
     public HttpInputOverHTTP(HttpChannelState state)
     {
         super(state);
     }
 
     @Override
-    protected void produceContent() throws IOException
+    public boolean addContent(Content content)
+    {
+        super.addContent(content);
+        return true;
```

Comment:
It is not immediately obvious why this is always return true here?  So at the very least it needs a comment to explain.

But if it is always return true because you have meta knowledge about when/why `addContent` is called, I'm still struggling to think of why `super.addContent` might return false and you would want to override.   Specifically if the state is not WAITING, then returning true here might cause a second thread to be dispatched to the channel

Reply:
This one always returns true because returning anything else in HTTP 1 could make the parser call `addContent` again right away which would end up adding two `Content` objects.

Regarding HTTP 2, the behavior strictly matches what the code was doing before the refactoring, so I didn't give it much thought as that was working as-is.

This is definitely one of those areas that will need more investigation/discussion.

Reply:
I'm VERY uncomfortable with using this to prevent multiple calls to addContent because the boolean return is a real meaningful value that is derived for an important state change in the HttpChannel

Consider we are asynchronous.... nobody has called `isReady()` yet but somebody calls `available()`, which ultimately causes `addContent` to be called, which changes the state of the `HttpChannelState` and returns a meaningful true/false as to weather a new thread needs to be scheduled or not.... that information is now lost.  

Let's think of the counter case?   When would a thread be in the parser when addContent would return false?   Well probably always if we go with the model that only the thread calling read is the one that runs the HttpInput machinery.... in which case perhaps it is no longer addContents job to make scheduling decisions???   Eitherway, this is a fundamental change that really needs to be thought out.

Reply:
So far, my gut feeling has been that the `addContent()` model is wrong and that `HttpInput` should pull data when it needs it instead of being pushed data whenever.

HTTP 1, 2 and FCGI are subtly different in this matter. We need to thoroughly investigate to better understand the reasons behind those differences.

Reply:
OK let's break this down:
+ H1 Blocking.  Read will be called by a single thread dispatched to a handler - or at least an application thread that is bound by a single dispatched thread.   `addContent` is always called by that thread in the scope of read.  If it blocks, then the onFillableCallback just wakes it up and it tries to fill and parse again.
+ H1 Async. Read can be called by any thread at any time, but we can say it is an app error if two threads call read at the same time.  `addContent` can be called either by a call to read  OR if `isReady()` has returned false, by the callback to onFillable calling HttpChannel.handle which calls produceRawContent - BUT not both as an application is not allowed to call read if isReady returned false.  The boolean return is again not a scheduling decision. Scheduling is done by the state change in HttpChannelState being checked in unhandle.
+ H2 Blocking. `addContent` might be called by a reading thread, only if demand(1) is implemented synchronously when there is available content.  If there is no content, the reading thread blocks and the callback to onContent eventually calls addContent which does the signal to wakeup the blocked thread.
+ H2 Async.  If there is content, then it acts like blocking. If there is no content, it is the callback to onContent that calls addContent and the boolean return is used to decide if a call to HttpChannel.handle is needed.

So for the four cases, we have four different scheduling mechanisms and behaviours!  Surely we can reduce this to two:
 + for blocking, any callbacks will do the unblock as soon as possible and let the reader do as much work as possible
 + for async, any callbacks should always enter HttpChannel.handle and let the producing logic there control what happens.

Reply:
Specifically if we get rid of the current H2 async behaviour then the boolean return to addContent is probably no longer necessary!   My lingering concern here is that rawContent may be added that in then entirely consumed by an interceptor, in which case blocking threads need to wakeup, run that consumption and then block again!